src/Http/Http.Api/src/MapActionExpressionTreeBuilder.cs

```diff
@@ -0,0 +1,385 @@
+                                           methodCall,
+                                           TargetArg,
+                                           HttpContextParameter);
+                    }
+                }
+                else if (method.ReturnType.IsGenericType &&
+                         method.ReturnType.GetGenericTypeDefinition() == typeof(ValueTask<>))
+                {
+                    var typeArg = method.ReturnType.GetGenericArguments()[0];
+
+                    if (typeof(IResult).IsAssignableFrom(typeArg))
+                    {
+                        body = Expression.Call(
+                                           ExecuteValueResultTaskOfTMethodInfo.MakeGenericMethod(typeArg),
+                                           methodCall,
+                                           TargetArg,
+                                           HttpContextParameter);
+                    }
+                    else
+                    {
+                        // ExecuteTask<T>(action(..), httpContext);
+                        body = Expression.Call(
+                                       ExecuteValueTaskOfTMethodInfo.MakeGenericMethod(typeArg),
+                                       methodCall,
+                                       TargetArg,
+                                       HttpContextParameter);
+                    }
+                }
+                else
+                {
+                    // TODO: Handle custom awaitables
+                    throw new NotSupportedException($"Unsupported return type: {method.ReturnType}");
+                }
+            }
+            else if (typeof(IResult).IsAssignableFrom(method.ReturnType))
+            {
+                body = Expression.Call(methodCall, ResultWriteResponseAsync, HttpContextParameter);
+            }
+            else if (method.ReturnType == typeof(string))
+            {
+                body = Expression.Call(StringResultWriteResponseAsync, HttpResponseExpr, methodCall, Expression.Constant(CancellationToken.None));
+            }
+            else
+            {
+                body = Expression.Call(JsonResultWriteResponseAsync, HttpResponseExpr, methodCall, Expression.Constant(CancellationToken.None));
+            }
+
+            Func<object?, HttpContext, ValueTask>? requestDelegate = null;
+
+            if (needBody)
+            {
+                // We need to generate the code for reading from the body before calling into the 
+                // delegate
+                var lambda = Expression.Lambda<Func<object?, HttpContext, object?, ValueTask>>(body, TargetArg, HttpContextParameter, DeserializedBodyArg);
+                var invoker = lambda.Compile();
+
+                requestDelegate = async (target, httpContext) =>
+                {
+                    var bodyValue = await httpContext.Request.ReadFromJsonAsync(bodyType!);
+
+                    await invoker(target, httpContext, bodyValue);
```

Comment:
We have to think about validation looks like here. The way this works in MVC is that deserialization errors and validation errors end up in model state. For ApiControllers a filter turns the model state in a 400 response (which includes the a serialized form of model state). 

* We could re-use ModelState here. It's not necessarily the best data structure to represent a graph of values, but MVC's validation system already understands it so that's one fewer thing to deal with.
* How do we surface the error to the user? We could bind a ModelState and let you deal with it. If we wanted to support MVC's auto validation as a feature (it's pretty neat and removes boiler plate), we could add it as a filter that's configured by default.

Reply:
We shouldn't add any implicit behavior until we figure out a way to remove said behavior. We currently don't have an application model so we need something reasonable that describes this behavior.

If we support something like model state, it would need to be an explicit parameter to the method (which is a much cleaner way of opting-into functionality).

Reply:
We're going to have some way to configure filters for these globally (`services.AddRouting(options => options.MapActionFilters. [Insert / Remove ])`). That should be sufficient to enable some implicit runtime behavior that's also removable.

But binding the type once we enable validation would be a good start.

Reply:
We should do that after, not in this PR

Reply:
Oh yeah, absolutely. Just pointing out that we have a gap here.

Reply:
> We're going to have some way to configure filters for these globally

Reply:
So you're going to end up with an application model, just without the "see and modify" part of it 😆